api/envoy/config/route/v3/route_components.proto

```diff
@@ -750,10 +750,49 @@ message RouteAction {
     ConnectConfig connect_config = 3;
   }
 
+  message ClusterSelector {
+    // Indicates the upstream cluster to which the request should be routed
+    // to.
+    string cluster = 1 [(validate.rules).string = {min_bytes: 1}];
+
+    // Envoy will determine the cluster to route to by reading the value of the
+    // HTTP header named by cluster_header from the request headers. If the
+    // header is not found or the referenced cluster does not exist, Envoy will
+    // return a 404 response.
+    //
+    // .. attention::
+    //
+    //   Internally, Envoy always uses the HTTP/2 *:authority* header to represent the HTTP/1
+    //   *Host* header. Thus, if attempting to match on *Host*, match on *:authority* instead.
+    string cluster_header = 2
+        [(validate.rules).string = {min_bytes: 1 well_known_regex: HTTP_HEADER_NAME strict: false}];
+
+    // Multiple upstream clusters can be specified for a given route. The
+    // request is routed to one of the upstream clusters based on weights
+    // assigned to each cluster. See
+    // :ref:`traffic splitting <config_http_conn_man_route_table_traffic_splitting_split>`
+    // for additional documentation.
+    WeightedCluster weighted_clusters = 3;
```

Comment:
I think if you involved weighted clusters in the repeated fallback, it gets pretty complicated. Do we need the full flexibility provided here?

Reply:
Yes, the use case we are introducing this for requires weighted clusters.

Can you please expand on the areas of complexity that you are concerned about? I did an initial pass at the implementation level implications of this change and did not stumble upon any major blockers, but I likely missed some important details.

Reply:
Note that we've discussed the possibility of moving WeightedClusters into an extensible LB policy framework, in which case it would not continue to live here anyway.  See discussion in #5598.

This probably won't happen soon enough to matter here; I mention it only as something to be consider in terms of how much effort we throw into the current implementation.

Reply:
SG, thanks for the pointer.

Reply:
@markdroth are you good with this API as currently specified for gRPC?

Reply:
I don't specifically see a problem w.r.t. WeightedClusters.  I do still have some questions about the overall approach here, but I'll comment on that below.